File: src/1274-Number-of-Ships-in-a-Rectangle/1274.java
class Solution {
    public int countShips(Sea sea, int[] tp, int[] bl) {
        if (bl[0] > tp[0] || bl[1] > tp[1] || !sea.hasShips(tp, bl)) return 0;
        if (tp[0] == bl[0] && tp[1] == bl[1]) return 1;
        int nx = (tp[0] + bl[0])/2, ny = (tp[1] + bl[1])/2;
        return countShips(sea, new int[]{nx, ny}, bl) + 
            countShips(sea, tp, new int[]{nx + 1, ny + 1}) + 
            countShips(sea, new int[]{nx, tp[1]}, new int[]{bl[0], ny + 1}) + 
            countShips(sea, new int[]{tp[0], ny}, new int[]{nx + 1, bl[1]});
    }
}

File: src/1274-Number-of-Ships-in-a-Rectangle/1274.py
class Solution(object):
    def countShips(self, sea: 'Sea', tp: 'Point', lp: 'Point') -> int:
        if lp.x > tp.x or lp.y > tp.y or not sea.hasShips(tp, lp):
            return 0
        if tp.x == lp.x and tp.y == lp.y:
            return 1
        nx, ny = (tp.x + lp.x)//2, (tp.y + lp.y)//2
        return self.countShips(sea, Point(nx, tp.y), Point(lp.x, ny + 1)) +\
            self.countShips(sea, Point(nx, ny), lp) + \
            self.countShips(sea, tp, Point(nx + 1, ny + 1)) + \
            self.countShips(sea, Point(tp.x, ny), Point(nx + 1, lp.y))